Add support for supplying generic method type parameters

vexx32 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Adds support for specifying type arguments for generic methods.

For example:

```ps1
[Array]::Empty[int]()
```

```ps1
$dictionary = [System.Collections.Concurrent.ConcurrentDictionary[string, int]]::new()
$dictionary.AddOrUpdate[float]($Key, $addEntryScript, $updateEntryScript, $FloatValue)
```

Many thanks to @dlwyatt for the initial parsing implementation! 😊 💖 

## PR Context

Resolves #5146

/cc @lzybkr @daxian-dbw 

/cc @TylerLeonhardt would this necessitate an update to language spec like the `dispose{}` one did?
This may affect tools not using an actual PS parser, but in general I think the impact will be minor overall, likely requiring some updates to any regex recognition patterns for some syntax highlighting rules.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/8533
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [x] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

PoshChan commented on build.psm1:
```diff
@@ -19,114 +19,96 @@ $tagsUpToDate = $false
 # Since code that uses Get-PSCommitID and Get-PSLatestTag assume that tags are fetched,
```
@vexx32, your last commit had 2 failures in `PowerShell-CI-linux`
Type extensible completion of type after using namespace.Command line: <using namespace System; [TextR>.IO.TextReader
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1: line 112
112:                         $expected.Found | Should -BeTrue
```
Type extensible completion of type after using namespace.Command line: <using namespace System.IO; [TextR>.TextReader
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1: line 112
112:                         $expected.Found | Should -BeTrue
```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-macos`
Type extensible completion of type after using namespace.Command line: <using namespace System; [TextR>.IO.TextReader
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/runner/runners/2.166.3/work/1/s/test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1: line 112
112:                         $expected.Found | Should -BeTrue
```
Type extensible completion of type after using namespace.Command line: <using namespace System.IO; [TextR>.TextReader
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/runner/runners/2.166.3/work/1/s/test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1: line 112
112:                         $expected.Found | Should -BeTrue
```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-windows`
Type extensible completion of type after using namespace.Command line: <using namespace System; [TextR>.IO.TextReader
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\ExtensibleCompletion.Tests.ps1: line 112
112:                         $expected.Found | Should -BeTrue
```
Type extensible completion of type after using namespace.Command line: <using namespace System.IO; [TextR>.TextReader
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\ExtensibleCompletion.Tests.ps1: line 112
112:                         $expected.Found | Should -BeTrue
```

TylerLeonhardt commented:
> This may affect tools not using an actual PS parser, but in general I think the impact will be minor overall, likely requiring some updates to any regex recognition patterns for some syntax highlighting rules.

Yeah this will possibly need a change to EditorSyntax.

iSazonov commented:
@vexx32 There are many formatting and style issue. Please fix issues for _changed_ code and move unrelated ones to another PR (I mean specially build.psm1 and some in other code). It will very help reviewers!

vexx32 commented:
@iSazonov ah, you're right. I'll have to do a bit of a rebase to clear that up, give me a few moments. I think a lot of those came from Dave's original code here, so I'll pare out what's needed and drop the unrelated style changes. Thanks! 💖

vexx32 committed:
:sparkles: Add parsing rules and AST members

- Add parsing rules to recognise generic method args

- Add AST members to MemberExpressionAst / InvokeMemberExpressionAst so
that the parsed information can be stored in the AST.

Co-authored-by: Dave Wyatt <dlwyatt115@gmail.com>

vexx32 committed:
:sparkles: Add member binding logic

- Adds member binding logic which works via invocation constraints

- Ensures that only an appropriate generic method can be bound if
generic type arguments are specified.

vexx32 committed:
:art: Add tab completion support

- Enables tab completion of types for generic method args

vexx32 committed:
:white_check_mark: Add tests for generic methods

vexx32 commented:
@iSazonov that should do it. Time for bed. 😁

PoshChan commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -2033,6 +2033,12 @@ private List<CompletionResult> GetResultForIdentifier(CompletionContext completi
             result = CompletionCompleters.CompleteCommandArgument(completionContext);
```
@vexx32, your last commit had 2 failures in `PowerShell-CI-linux`
Type extensible completion of type after using namespace.Command line: <using namespace System; [TextR>.IO.TextReader
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1: line 112
112:                         $expected.Found | Should -BeTrue
```
Type extensible completion of type after using namespace.Command line: <using namespace System.IO; [TextR>.TextReader
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1: line 112
112:                         $expected.Found | Should -BeTrue
```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-macos`
Type extensible completion of type after using namespace.Command line: <using namespace System; [TextR>.IO.TextReader
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/runner/runners/2.166.3/work/1/s/test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1: line 112
112:                         $expected.Found | Should -BeTrue
```
Type extensible completion of type after using namespace.Command line: <using namespace System.IO; [TextR>.TextReader
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/runner/runners/2.166.3/work/1/s/test/powershell/Language/Parser/ExtensibleCompletion.Tests.ps1: line 112
112:                         $expected.Found | Should -BeTrue
```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-windows`
Type extensible completion of type after using namespace.Command line: <using namespace System; [TextR>.IO.TextReader
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\ExtensibleCompletion.Tests.ps1: line 112
112:                         $expected.Found | Should -BeTrue
```
Type extensible completion of type after using namespace.Command line: <using namespace System.IO; [TextR>.TextReader
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Parser\ExtensibleCompletion.Tests.ps1: line 112
112:                         $expected.Found | Should -BeTrue
```

vexx32 committed:
:wrench: Narrow completion to MemberExpressionAst

PoshChan commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -2033,6 +2033,14 @@ private List<CompletionResult> GetResultForIdentifier(CompletionContext completi
             result = CompletionCompleters.CompleteCommandArgument(completionContext);
```
@vexx32, your last commit had 5 failures in `PowerShell-CI-static-analysis`
Verify Markdown Links.Verify links in /home/vsts/work/1/s/CHANGELOG/6.0.md.https://blogs.msdn.microsoft.com/powershell/2017/07/14/powershell-6-0-roadmap-coreclr-backwards-compatibility-and-more/ should work
```powershell
Failed to complete request to "https://blogs.msdn.microsoft.com/powershell/2017/07/14/powershell-6-0-roadmap-coreclr-backwards-compatibility-and-more/". Response status code does not indicate success: 404 (Not Found).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 119
119:                                     throw "Failed to complete request to `"$url`". $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/CHANGELOG/6.0.md.https://blogs.msdn.microsoft.com/dotnet/2016/09/26/introducing-net-standard/ should work
```powershell
Failed to complete request to "https://blogs.msdn.microsoft.com/dotnet/2016/09/26/introducing-net-standard/". Response status code does not indicate success: 404 (Not Found).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 119
119:                                     throw "Failed to complete request to `"$url`". $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/CHANGELOG/6.0.md.https://blogs.msdn.microsoft.com/powershell/2017/01/31/powershell-open-source-community-dashboard/ should work
```powershell
Failed to complete request to "https://blogs.msdn.microsoft.com/powershell/2017/01/31/powershell-open-source-community-dashboard/". Response status code does not indicate success: 404 (Not Found).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 119
119:                                     throw "Failed to complete request to `"$url`". $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/demos/WindowsPowerShellModules/README.md.https://blogs.msdn.microsoft.com/powershell/2017/07/14/powershell-6-0-roadmap-coreclr-backwards-compatibility-and-more/ should work
```powershell
Failed to complete request to "https://blogs.msdn.microsoft.com/powershell/2017/07/14/powershell-6-0-roadmap-coreclr-backwards-compatibility-and-more/". Response status code does not indicate success: 404 (Not Found).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 119
119:                                     throw "Failed to complete request to `"$url`". $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/README.md.https://blogs.msdn.microsoft.com/powershell/2017/01/31/powershell-open-source-community-dashboard/ should work
```powershell
Failed to complete request to "https://blogs.msdn.microsoft.com/powershell/2017/01/31/powershell-open-source-community-dashboard/". Response status code does not indicate success: 404 (Not Found).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 119
119:                                     throw "Failed to complete request to `"$url`". $($_.Exception.Message)"

```

vexx32 replied:
Unrelated to this PR. I blame Travis for irresponsibly including Markdown links that don't exist yet. 😝

adityapatwardhan commented:
@rjmholt Could you review this please?

rjmholt commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -7678,27 +7690,96 @@ private ExpressionAst MemberAccessRule(ExpressionAst targetExpr, Token operatorT
+            Token lBracket = NextToken();
+            rBracketToken = null;
+
+            if (lBracket.Kind == TokenKind.LBracket)
```
If you invert this condition, we can dispense with an indentation level. If we call `NextToken()` and the next token is not an LBracket, it sounds like we might need to return the token and indicate that this method does not apply?

vexx32 replied:
Returning null just leaves the `MemberAccessRule` with a null value for method type arguments. Since it's impossible to determine the difference between a property and a method in the parser until we get the final `()` token at the end, calling the method here and getting null is OK. The `MemberAccessRule` only needs to know if there were generic method arguments or not and to pass that information along accordingly.

If there were none, a null result is expected and just means the `MemberAccessRule` and the resulting AST element has no generic type arguments. 

Good point on the indent level, though, we can avoid needing to `Resync` here by inverting the conditional and using `PeekToken()` as well which is a nice plus! 👍

vexx32 replied:
I lied, we can't avoid a Resync() unfortunately. It is still simpler to code the way you suggested though.

Left a detailed comment in the code explaining why PeekToken() can't be used in that specific place without breaking everything.

rjmholt commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -7678,27 +7690,96 @@ private ExpressionAst MemberAccessRule(ExpressionAst targetExpr, Token operatorT
+            {
+                // This is either a member expression with generic type arguments, or some sort of collection index
+                // on a property.
+                var oldTokenizerMode = _tokenizer.Mode;
```
Please use explicit type here

rjmholt commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -7678,27 +7690,96 @@ private ExpressionAst MemberAccessRule(ExpressionAst targetExpr, Token operatorT
+
+                    SkipNewlines();
+                    Token firstToken = NextToken();
+                    if (firstToken.Kind != TokenKind.Number
+                        && (firstToken.Kind == TokenKind.Identifier || firstToken.Kind == TokenKind.LBracket))
```
The logic here seems to test that the token kind is not one thing and then is one of two others. The second conditions should preclude the first, no?

One question is what unexpected tokens do here and how we signal that to the caller

rjmholt replied:
If we can identify some variations on the logic here, we should ideally add tests for each possibility so that we have a strong understanding of the behaviour here

vexx32 replied:
So I could be misremembering, but I think `Identifier` _includes_ `Number` (it's a named flag combination, if I recall correctly?)

Will have to look back over this code to answer your other point. 🙂

vexx32 replied:
Re-checked this, and yeah you're right, the first condition isn't needed. I must have been thinking of something else, no [Flags] shenanigans here. ^^

As for what unexpected tokens do here.... not sure exactly what you mean. Looking at the code, we use ReportIncompleteInput() but beyond that I'm not quite sure. 🙂

rjmholt commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -7678,27 +7690,96 @@ private ExpressionAst MemberAccessRule(ExpressionAst targetExpr, Token operatorT
+                        && (firstToken.Kind == TokenKind.Identifier || firstToken.Kind == TokenKind.LBracket))
+                    {
+                        resyncIndex = -1;
+                        genericTypes = GenericArgumentsRule(firstToken, false, out rBracketToken);
```
Please label the `false` parameter

rjmholt commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -1460,30 +1459,41 @@ private ITypeName GenericTypeArgumentsRule(Token genericTypeName, Token firstTok
+            return genericArguments;
+        }
+
+        private ITypeName GenericTypeArgumentsRule(Token genericTypeName, Token firstToken, bool unBracketedGenericArg)
```
Given that we're breaking these methods out into separate pieces, we should try to rename them to refer to exactly what they parse. I'm not clear on the difference between `GenericArgumentsRule` and `GenericTypeArgumentsRule`

vexx32 replied:
The first should probably be `GenericMethodArgumentsRule`, but I'll double check.

vexx32 replied:
Ah, I see where the confusion comes from. Okay... the current `GenericTypeArgumentsRule` returns a `GenericTypeName`, so I'm going to rename:

- `GenericTypeArgumentsRule` => `GenericTypeNameRule`
- `GenericArgumentsRule` => `GenericTypeArgumentsRule`

That seems to make the most sense to me.

iSazonov replied:
> GenericArgumentsRule => GenericTypeArgumentsRule

This can mislead if anybody looks a commit history. Could we use another name for new `GenericTypeArgumentsRule`?

vexx32 replied:
I'm open to suggestions if the names can be clearer, but the naming of the original method this is built out of puts us in a little bit of a corner in that way.

rjmholt commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -7891,14 +7909,26 @@ public MemberExpressionAst(IScriptExtent extent, ExpressionAst expression, Comma
+        /// <summary>
+        /// List of generic type arguments passed to this member.
+        /// </summary>
+        public ReadOnlyCollection<ITypeName> GenericTypeArguments { get; private set; }
```
Is the `private` setter used anywhere?

vexx32 replied:
Doesn't appear to be used anymore, I'll take it out. Think it was being set after construction in an earlier version of this code.

rjmholt commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -7678,27 +7690,96 @@ private ExpressionAst MemberAccessRule(ExpressionAst targetExpr, Token operatorT
+                        if (rBracketToken.Kind != TokenKind.RBracket)
+                        {
+                            UngetToken(rBracketToken);
+                            ReportIncompleteInput(
```
If we can, we should test for this case

vexx32 replied:
Absolutely! Not sure how to test for this, though. 😅

vexx32 commented on src/System.Management.Automation/engine/MshMemberInfo.cs:
```diff
@@ -1997,6 +2014,7 @@ public override int GetHashCode()
 
                 result = result * 397 + (MethodTargetType != null ? MethodTargetType.GetHashCode() : 0);
                 result = result * 397 + ParameterTypes.SequenceGetHashCode();
+                result = result * 397 + GenericTypeParameters.SequenceGetHashCode();
```
Might be a good idea to open an issue for that. I can do that for this segment in this PR if you'd like? 🙂

vexx32 commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -6281,18 +6287,28 @@ public object VisitMemberExpression(MemberExpressionAst memberExpressionAst)
         {
-            var arguments = invokeMemberExpressionAst.Arguments;
+            var arguments = invokeMemberExpressionAst.Arguments
+                ?.Select(Compiler.GetTypeConstraintForMethodResolution)
+                .ToArray();
```
I copied the original pattern used a couple lines above for the method argument types here, but you're right. I'll move both that and this one to using a more traditional loop.

vexx32 committed:
:recycle: Address review comments

vexx32 committed:
:white_check_mark: Add tests for incomplete input

iSazonov commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -2033,6 +2033,14 @@ private List<CompletionResult> GetResultForIdentifier(CompletionContext completi
             replacementIndex = completionContext.ReplacementIndex;
             replacementLength = completionContext.ReplacementLength;
+
+            if (result.Count == 0
+                && lastAst?.Find(a => a is MemberExpressionAst, searchNestedScriptBlocks: false) != null
+                && completionContext.TokenAtCursor.TokenFlags.HasFlag(TokenFlags.TypeName))
```
Maybe swap 2 and 3 conditions?

iSazonov commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -1433,7 +1438,24 @@ private static bool IsInvocationConstraintSatisfied(OverloadCandidate overloadCa
                         }
                     }
 
+                    if (genericParameters?.Length > 0 && method.method is MethodInfo originalMethod)
```
It seems `genericParameters` is never null.

iSazonov commented on src/System.Management.Automation/engine/MshMemberInfo.cs:
```diff
@@ -1905,9 +1905,18 @@ internal class PSMethodInvocationConstraints
+        {
+            MethodTargetType = methodTargetType;
             _parameterTypes = parameterTypes;
+            GenericTypeParameters = genericTypeParameters?.ToArray() ?? Array.Empty<Type>();
```
Why do we need `ToArray()`?

vexx32 replied:
We don't anymore. I probably forgot to remove it :)

iSazonov commented on src/System.Management.Automation/engine/MshMemberInfo.cs:
```diff
@@ -1922,6 +1931,11 @@ internal class PSMethodInvocationConstraints
+        /// <summary>
+        /// Gets the generic type parameters for the method invocation.
+        /// </summary>
+        public Type[] GenericTypeParameters { get; private set; }
```
Do we need `private set`?

vexx32 replied:
Nope, doesn't look like it. Nice catch!

iSazonov commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -1159,24 +1159,30 @@ internal static Type GetTypeConstraintForMethodResolution(ExpressionAst expr)
+            Type[] genericArguments = null)
         {
-            if (targetType == null && argType == null)
+            if (targetType == null && argType == null && !(genericArguments?.Length > 0))
```
Could you please make this more readable and supportable:
```suggestion
            if (targetType == null && argType == null && (genericArguments == null || genericArguments .Length == 0))
```

iSazonov commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -1159,24 +1159,30 @@ internal static Type GetTypeConstraintForMethodResolution(ExpressionAst expr)
+            Type[] genericArguments = null)
         {
-            if (targetType == null && (argTypes == null || argTypes.Length == 0))
+            if (targetType == null && !(argTypes?.Length > 0) && !(genericArguments?.Length > 0))
```
The same. Please make this more readable.

iSazonov commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -6300,18 +6306,57 @@ public object VisitMemberExpression(MemberExpressionAst memberExpressionAst)
         {
-            var arguments = invokeMemberExpressionAst.Arguments;
+            Type[] argumentTypes = null;
+            if (invokeMemberExpressionAst.Arguments != null)
+            {
+                argumentTypes = new Type[invokeMemberExpressionAst.Arguments.Count];
+                for (var i = 0; i < invokeMemberExpressionAst.Arguments.Count; i++)
+                {
+                    argumentTypes[i] = GetTypeConstraintForMethodResolution(invokeMemberExpressionAst.Arguments[i]);
+                }
+            }
```
I suggest to put `invokeMemberExpressionAst.Arguments` in local variable because it is a compiler and performance is important.

Please use the pattern below too.

iSazonov commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -7863,6 +7869,11 @@ public MemberExpressionAst(IScriptExtent extent, ExpressionAst expression, Comma
             SetParent(member);
             this.Static = @static;
+
+            if (genericTypes?.Any() == true)
+            {
+                this.GenericTypeArguments = new ReadOnlyCollection<ITypeName>(genericTypes.ToArray());
```
Can we avoid these LINQ?
Maybe always initialize GenericTypeArguments with empty collection if right argument is null? This allows us remove many null checks.

iSazonov commented on test/powershell/Language/Parser/MethodInvocation.Tests.ps1:
```diff
@@ -81,3 +81,110 @@ namespace MSFT_716893
+        [System.Linq.Enumerable]::Select[int, int](
+            [int[]](0..10),
+            [func[int, int]] { $args[0] + 2 }
+        ) | Should -Be @(2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)
```
It seems we need `Compare-Object -SyncWindow 0` to compare arrays.

vexx32 replied:
In what way? `Should` will compare arrays and ensure the order is the same.

iSazonov replied:
Doesn't a pipeline enumerate?

vexx32 replied:
Yep. `Should` does some extra handling to let it compare arrays from the pipeline. I'm not sure of the implementation details, but `1, 2, 3 | Should -Be 2, 3, 1` will fail, where `1, 2, 3 | Should -Be 1, 2, 3` will pass.

```
PS> 1, 2, 3 | Should -Be 2, 3, 1 
InvalidResult: 
Line |
   1 | 1, 2, 3 | Should -Be 2, 3, 1
     | Expected @(2, 3, 1), but got @(1, 2, 3).

PS> 1, 2, 3 | Should -Be 1, 2, 3
```

vexx32 commented:
@rjmholt please update your review when you can. Thanks! 😊 💖 

@daxian-dbw did you want to review this as well? ^^

adityapatwardhan commented:
@rjmholt @daxian-dbw - Do we need a RFC for this?

adityapatwardhan commented:
@vexx32 Please resolve merge conflicts

vexx32 commented:
@adityapatwardhan all sorted

rjmholt commented:
@PoshChan please remind me in 2 days

iSazonov commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -1158,24 +1158,34 @@ internal static Type GetTypeConstraintForMethodResolution(ExpressionAst expr)
+            Type[] genericArguments = null)
         {
-            if (targetType == null && argType == null)
+            if (targetType == null
+                && argType == null
+                && (genericArguments == null || genericArguments.Length == 0))
```
```suggestion
            if (targetType is null
                && argType is null
                && (genericArguments is null || genericArguments.Length == 0))
```

iSazonov commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -1158,24 +1158,34 @@ internal static Type GetTypeConstraintForMethodResolution(ExpressionAst expr)
+            Type[] genericArguments = null)
         {
-            if (targetType == null && (argTypes == null || argTypes.Length == 0))
+            if (targetType == null
+                && (argTypes == null || argTypes.Length == 0)
+                && (genericArguments == null || genericArguments.Length == 0))
```
```suggestion
            if (targetType is null
                && (argTypes is null || argTypes.Length == 0)
                && (genericArguments is null || genericArguments.Length == 0))
```

iSazonov commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -6299,18 +6309,59 @@ public object VisitMemberExpression(MemberExpressionAst memberExpressionAst)
-            var arguments = invokeMemberExpressionAst.Arguments;
+            ReadOnlyCollection<ExpressionAst> arguments = invokeMemberExpressionAst.Arguments;
+            Type[] argumentTypes = null;
+            if (arguments != null)
```
```suggestion
            if (arguments is not null)
```

iSazonov commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -6299,18 +6309,59 @@ public object VisitMemberExpression(MemberExpressionAst memberExpressionAst)
+
+            ReadOnlyCollection<ITypeName> genericArguments = invokeMemberExpressionAst.GenericTypeArguments;
+            Type[] genericTypeArguments = null;
+            if (genericArguments != null)
```
```suggestion
            if (genericArguments is not null)
```

iSazonov commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -6299,18 +6309,59 @@ public object VisitMemberExpression(MemberExpressionAst memberExpressionAst)
             Type targetTypeConstraint = null;
-            var arguments = invokeMemberExpressionAst.Arguments;
+            Type[] argumentTypes = null;
+            if (invokeMemberExpressionAst.Arguments != null)
```
```suggestion
            if (invokeMemberExpressionAst.Arguments is not null)
```

iSazonov commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -2033,6 +2033,14 @@ private List<CompletionResult> GetResultForIdentifier(CompletionContext completi
+
+            if (result.Count == 0
+                && completionContext.TokenAtCursor.TokenFlags.HasFlag(TokenFlags.TypeName)
+                && lastAst?.Find(a => a is MemberExpressionAst, searchNestedScriptBlocks: false) != null)
```
```suggestion
                && lastAst?.Find(a => a is MemberExpressionAst, searchNestedScriptBlocks: false) is not null)
```

iSazonov commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -1394,32 +1394,37 @@ private static bool IsInvocationConstraintSatisfied(OverloadCandidate overloadCa
                 // generic methods need to be double checked in a loop below - generic methods can be rejected if type inference fails
-                (methods[0].parameters.Length == arguments.Length))
+                && !methods[0].isGeneric
+                && (methods[0].method == null || !(methods[0].method.DeclaringType.IsGenericTypeDefinition))
```
```suggestion
                && (methods[0].method is null || !(methods[0].method.DeclaringType.IsGenericTypeDefinition))
```

iSazonov commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -1394,32 +1394,37 @@ private static bool IsInvocationConstraintSatisfied(OverloadCandidate overloadCa
                 MethodInformation method = methods[i];
 
-                if (method.method != null && method.method.DeclaringType.IsGenericTypeDefinition)
+                if (method.method != null && method.method.DeclaringType.IsGenericTypeDefinition
```
```suggestion
                if (method.method is not null && method.method.DeclaringType.IsGenericTypeDefinition
```

vexx32 commented:
Linux elevated test on last commit failed with a seemingly unrelated error. Restarting it to see if it's sporadic, this is the 2nd time I've come across it (across different PRs, too, I think).

vexx32 commented:
not sure what's up with codeQL here... the log shows nothing, and I get no errors building locally.

PoshChan commented:
@rjmholt, this is the reminder you requested 2 days ago

adityapatwardhan commented on test/powershell/Language/Parser/MethodInvocation.Tests.ps1:
```diff
@@ -81,3 +81,110 @@ namespace MSFT_716893
     }
 }
+
+Describe 'Generic Method invocation' {
```
```suggestion
Describe 'Generic Method invocation' -Tags "CI" {
```

adityapatwardhan commented:
@vexx32 Also can you have a look at the CodeFactor issues?

vexx32 commented:
@adityapatwardhan that should sort it out. Some of the codefactor things were more or less tangential, but I figured they were small enough to fix anyway, save someone else the hassle later.

iSazonov commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -2102,6 +2102,14 @@ private List<CompletionResult> GetResultForIdentifier(CompletionContext completi
             replacementIndex = completionContext.ReplacementIndex;
             replacementLength = completionContext.ReplacementLength;
+
+            if (result.Count == 0
+                && completionContext.TokenAtCursor.TokenFlags.HasFlag(TokenFlags.TypeName)
+                && lastAst?.Find(a => a is MemberExpressionAst, searchNestedScriptBlocks: false) is not null)
+            {
+                result = CompletionCompleters.CompleteType(completionContext.TokenAtCursor.Text).ToList();
+            }
+
```
I don't see new tests for the code. What is a scenario we address here?

vexx32 replied:
I think this was something like `[array]::empty[str`<tab> should complete to types known for `string`.

iSazonov replied:
You can move this to follow PR to reduce that PR.

vexx32 replied:
I don't think it makes that much sense to separate this out tbh.

iSazonov replied:
You know our experience :-) - the more compact change and the less risky the faster we can review and merge.
Since you do not prepare tests for the scenario I believe we can address the scenario in follow PR and fast merge since there is small risk.

iSazonov reviewed (commented):
@vexx32 Please move all style changes in another PR (I will merge it before). It would save a lot of our time on code review.

vexx32 commented:
I will have a look at separating out purely style changes when I can, although I'm not sure I made that many in this PR.

rjmholt commented on test/powershell/Language/Parser/MethodInvocation.Tests.ps1:
```diff
@@ -81,3 +81,110 @@ namespace MSFT_716893
     }
 }
+
+Describe 'Generic Method invocation' -Tags 'CI' {
```
A few other test suggestions:

- The big one for me is different kinds of error
    - `[array]::empty[notarealtype]()`
    - `[array]::empty[type]]()`
    - `[array]::empty[type, type]()`
    - `$object.Method[type,]()`
    - `$object.Method[]()`
    - `$object.Method[,]()`
    - `$object.Method[,type]()`
    - `$object.Method[type()`
    - `$object.Method[type)`
- Also tests around non-invoked methods (the method group info result):
    - `$object.Method[type]`
    - `$object.Method[type, type]`
- And something like `[Enumerable]::Select[List[int]](...)`

vexx32 replied:
> Also tests around non-invoked methods (the method group info result):
> * $object.Method[type]
> * $object.Method[type, type]

Hmm, that's a good point. The logic that retrieves method group info happens before (and separately to) the logic that actually resolves the method. I will double check this evening, but from what I recall, there isn't an easy way to pull the resolution logic into the method group result.

It might be doable, but I'm unsure if it is worth the effort in this instance. Currently I think the results from `$object.Method` and `$object.Method[...types...]` are identical, the resolution is only done when it's actually invoked at the moment. 

I'll look to add some more tests along those lines as well though, thanks!

rjmholt replied:
> Currently I think the results from $object.Method and $object.Method[...types...] are identical, the resolution is only done when it's actually invoked at the moment.

Yeah, I can see that making sense. The important part, I think, is that we capture what the expected behaviour is with tests.

vexx32 replied:
I've added test cases for all the variants you mentioned. I think they're all behaving in a pretty consistent way. Had to go in and figure out how to give some better runtime errors for cases where a method overload can't be found / the type(s) specified don't exist, and added some tests for those cases too.

rjmholt commented:
This mostly looks good to me, just want a few more tests

vexx32 committed:
:white_check_mark: Additional tests

vexx32 commented on test/powershell/Language/Parser/MethodInvocation.Tests.ps1:
```diff
@@ -81,3 +81,196 @@ namespace MSFT_716893
+        ) | Should -Be @(2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)
+    }
+
+    It 'gives a runtime error if too many type parameters are given' {
+        try {
+            [array]::Empty[string, int]()
+        }
+        catch {
+            $_.Exception.Message | Should -BeExactly 'Cannot find an overload for "Empty" and the argument count: "0".'
+            $_.FullyQualifiedErrorId | Should -BeExactly 'MethodCountCouldNotFindBest'
+        }
+    }
+
+    It 'gives a runtime error if a nonexistent type is specified' {
+        try {
+            [array]::Empty[thisdoesnotexist]()
+        }
+        catch {
+            $_.Exception.Message | Should -BeExactly 'Cannot find an overload for "empty" and the argument count: "0".'
+            $_.FullyQualifiedErrorId | Should -BeExactly 'MethodCountCouldNotFindBest'
+        }
+    }
```
@rjmholt these two are the ones I like the least. I'd like there to be more actionable errors here, such as:

- `type specified for generic method was not found`
- `a valid generic method overload that takes the specified number of type arguments was not found`

Thus far though, I've not had much luck with incorporating logic for this kind of case. 😕

vexx32 replied:
I managed to find some... okay... solutions here. Part of the problem getting a reasonable error message for something like `[array]::Empty[notexistent]()` is that a nonexistent type name doesn't leave retain its original lexeme at this point in the invocation, we just have a `null` entry to work with.

This is still miles better than the default error message that shows up here, though.

vexx32 committed:
:recycle: Address review comments

vexx32 committed:
:recycle: Use more descriptive error messages

The default error message for failing to find a resolvable method here
is very unhelpful for some common oops cases with generic methods.
Added some extra error messages that can turn up here for better UX.

vexx32 commented on src/System.Management.Automation/engine/MshMemberInfo.cs:
```diff
@@ -2000,8 +2017,9 @@ public override int GetHashCode()
+                result = (result * 397) + ParameterTypes.SequenceGetHashCode();
+                result = (result * 397) + GenericTypeParameters.SequenceGetHashCode();
 
                 return result;
```
Done!

vexx32 commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -1394,32 +1394,37 @@ private static bool IsInvocationConstraintSatisfied(OverloadCandidate overloadCa
                     Array.Copy(argumentTypes, argumentTypesForTypeInference, argumentTypes.Length);
-                    if (invocationConstraints != null && invocationConstraints.ParameterTypes != null)
+
+                    if (invocationConstraints?.ParameterTypes != null)
```
Thanks, added this one too. :)

rjmholt commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -7873,10 +7873,16 @@ public class MemberExpressionAst : ExpressionAst, ISupportsAssignment
+            ExpressionAst expression,
+            CommandElementAst member,
+            bool @static,
+            IList<ITypeName> genericTypes = null)
```
Just saw this -- [we should probably not have default values in public APIs](https://rules.sonarsource.com/csharp/RSPEC-2360)

rjmholt commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -7902,11 +7909,18 @@ public MemberExpressionAst(IScriptExtent extent, ExpressionAst expression, Comma
+            CommandElementAst member,
+            bool @static,
+            bool nullConditional,
+            IList<ITypeName> genericTypes = null)
```
Ditto

rjmholt commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -7986,11 +8012,18 @@ public class InvokeMemberExpressionAst : MemberExpressionAst, ISupportsAssignmen
+            CommandElementAst method,
+            IEnumerable<ExpressionAst> arguments,
+            bool @static,
+            IList<ITypeName> genericTypes = null)
```
Ditto

rjmholt commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -8013,11 +8046,19 @@ public InvokeMemberExpressionAst(IScriptExtent extent, ExpressionAst expression,
+            IEnumerable<ExpressionAst> arguments,
+            bool @static,
+            bool nullConditional,
+            IList<ITypeName> genericTypes = null)
```
Ditto

vexx32 committed:
:recycle: Add constructor overloads

To avoid using default values in public API methods, added some extra
constructor overloads for the relevant AST types here.

vexx32 commented:
@rjmholt I've removed the default values for the public API constructors and added extra overloads instead, thanks for pointing that out! 💖

rjmholt commented:
@adityapatwardhan please refresh your review, @daxian-dbw please take a look when you get time

iSazonov commented:
I added Breaking-Change label to notice EditorSyntax.

rjmholt commented:
Just a note: since 7.2 is going to RC soon, we're intending to merge this after that so it gets some time in the 7.3 preview

SeeminglyScience commented:
@daxian-dbw @adityapatwardhan Any chance this PR can make it into the next preview? 🤞

iSazonov commented:
We need @andschwa review too (for possible PSES breaking).

daxian-dbw commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -1394,32 +1394,37 @@ private static bool IsInvocationConstraintSatisfied(OverloadCandidate overloadCa
             }
 
             Type[] argumentTypes = arguments.Select(EffectiveArgumentType).ToArray();
+            Type[] genericParameters = invocationConstraints?.GenericTypeParameters ?? Array.Empty<Type>();
```
The `?? Array.Empty<Type>();` part is not needed, as your changes to `PSMethodInvocationConstraints` constructor already makes sure it's always initialized with an empty array in case it's null.
But I don't know, maybe we should keep it here and remove it from the constructor.

daxian-dbw replied:
> I think it's possible for invocationConstraints itself here to be null

Yeah, I agree. Then let's remove it from the constructor.

daxian-dbw replied:
Addressed.

daxian-dbw commented on src/System.Management.Automation/engine/MshMemberInfo.cs:
```diff
@@ -1909,9 +1909,18 @@ internal class PSMethodInvocationConstraints
+        {
+            MethodTargetType = methodTargetType;
             _parameterTypes = parameterTypes;
+            GenericTypeParameters = genericTypeParameters ?? Array.Empty<Type>();
```
Maybe this should just be `GenericTypeParameters = genericTypeParameters;`, just like how `_parameterTypes` is handled.

daxian-dbw replied:
Addressed

vexx32 commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -1590,6 +1611,22 @@ private static bool IsInvocationConstraintSatisfied(OverloadCandidate overloadCa
                         methods[0].method.DeclaringType.FullName);
                     return null;
                 }
+                else if (genericParameters.Length != 0 && genericParameters.Contains(null))
```
I'd have to take another look... I have a recollection that I tried to do something like this and it didn't pan out.

daxian-dbw replied:
Tracking issue #16809 was opened.

vexx32 commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -6310,18 +6320,59 @@ public object VisitMemberExpression(MemberExpressionAst memberExpressionAst)
+
+                for (var i = 0; i < genericArguments.Count; i++)
+                {
+                    genericTypeArguments[i] = genericArguments[i].GetReflectionType();
```
Is this where you were suggesting we handle the null case and throw some kind of error? I think initially I wasn't sure we would want to throw from here, but it looks like it should be safe to emit some kind of error here... although I don't really know how best to do that, is it OK to throw exceptions from here, or will that cause issues for other things?

daxian-dbw replied:
Looked more into it, and I think we got a problem in constructing `PSMethodInvocationConstraints`:
- Using `GetReflectionType()` only means we only tries to resolve the argument types (generic argument or method invocation argument) at compilation time. It doesn't handle the case where the type gets loaded in at runtime.

We should do something similar to [CompileTypeName](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/parser/Compiler.cs#L6243) -- try `GetReflectionType()` first and if that returns null, fall back to [TypeOps.ResolveTypeName](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs#L2116) in `FindBestMethodImpl`.

I think that should be done in a separate PR. I will open an issue about this.

daxian-dbw replied:
Tracking issue #16809 was opened.

vexx32 commented on src/System.Management.Automation/engine/parser/Compiler.cs:
```diff
@@ -6310,18 +6320,59 @@ public object VisitMemberExpression(MemberExpressionAst memberExpressionAst)
+
             TypeDefinitionAst typeDefinitionAst = Ast.GetAncestorTypeDefinitionAst(invokeMemberExpressionAst);
+
+            Type[] genericArguments = null;
+            if (invokeMemberExpressionAst.GenericTypeArguments.Count > 0)
+            {
+                genericArguments = new Type[invokeMemberExpressionAst.GenericTypeArguments.Count];
+                for (var i = 0; i < invokeMemberExpressionAst.GenericTypeArguments.Count; i++)
+                {
+                    genericArguments[i] = invokeMemberExpressionAst.GenericTypeArguments[i].GetReflectionType();
+                }
+            }
```
Oh, huh, this is just for constructors, then? Aight, we can remove this change if that's the case.

daxian-dbw replied:
Yes, this is just for calling base constructor. The constructor for `BaseCtorInvokeMemberExpressionAst` doesn't access any generic arguments, so we should remove this.

daxian-dbw replied:
Addressed.

vexx32 commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -1437,7 +1437,7 @@ private ITypeName GetSingleGenericArgument(Token firstToken)
         }
 
-        private ITypeName GenericTypeArgumentsRule(Token genericTypeName, Token firstToken, bool unBracketedGenericArg)
+        private List<ITypeName> GenericTypeArgumentsRule(Token firstToken, bool unbracketedGenericArg, out Token lastToken)
```
Yeah, I don't think it was before this change either. Wasn't sure if it was worth removing in this PR or better left to a later cleanup PR or something.

daxian-dbw replied:
It was used by the original `GenericTypeArgumentsRule` before this change. The old `GenericTypeArgumentsRule` method was split into 2 methods in this PR and that made this new `GenericTypeArgumentsRule` no longer use this parameter.

daxian-dbw replied:
Addressed in my latest commit.

vexx32 commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -1446,13 +1446,12 @@ private ITypeName GenericTypeArgumentsRule(Token genericTypeName, Token firstTok
             genericArguments.Add(typeName);
 
-            Token commaOrRBracketToken;
             Token token;
```
Makes sense, aye.

vexx32 commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -1467,43 +1466,56 @@ private ITypeName GenericTypeArgumentsRule(Token genericTypeName, Token firstTok
+                UngetToken(rBracketToken);
+                ReportIncompleteInput(
+                    Before(rBracketToken),
+                    nameof(ParserStrings.UnexpectedToken),
+                    ParserStrings.UnexpectedToken,
```
No recollection whatsoever, I'll put it back.

vexx32 commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -7715,34 +7729,107 @@ private ExpressionAst MemberAccessRule(ExpressionAst targetExpr, Token operatorT
+                // We cannot avoid this Resync(); if we use PeekToken() to try to avoid a Resync(), the method called
+                // after this (NextInvokeMemberToken()) will note that an _ungotToken is present and assume an error
+                // state. That will cause any property accesses or non-generic method calls to throw a parse error.
+                Resync(resyncIndex);
```
If I understand correctly it should be OK to just return null here and allow the `NextInvokeMemberToken()` method to error? Or should we explicitly emit an error from here in this case?

vexx32 replied:
And do you think we can/should avoid the later resync in this method as well?

daxian-dbw replied:
Sorry, I was wrong about the above suggestion. We indeed cannot avoid `resync()` because `NextMemberAccessToken()` also assumes `_ungotToken == null`.

vexx32 committed:
Address review comments

daxian-dbw committed:
Fix the build break

daxian-dbw commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -7715,34 +7729,107 @@ private ExpressionAst MemberAccessRule(ExpressionAst targetExpr, Token operatorT
+                // Member name may be an incomplete token like `$a.$(Command-Name`; we do not look for generic args or
+                // invocation token(s) if the member name token is recognisably incomplete.
+                genericTypeArguments = GenericMethodArgumentsRule(out rBracket);
                 Token lParen = NextInvokeMemberToken();
```
I was wrong about this suggestion. It turns out we indeed cannot avoid `resync` because `NextMemberAccessToken()` also assumes `_ungotToken == null`. I have reverted this change in my commit.

daxian-dbw commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -7715,34 +7729,107 @@ private ExpressionAst MemberAccessRule(ExpressionAst targetExpr, Token operatorT
+                    // Fall through to the last available token expected in a member name for calculating the expected
+                    // position of the lParen. This accounts for things like `[Array]::Empty[int[]()` so that we don't
+                    // get unhandled behaviour when input is expected but incomplete.
+                    int endColumnNumber = rBracket?.Extent?.EndColumnNumber
```
This was updated in my commit. Also updated the comments.

daxian-dbw commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -1446,64 +1446,75 @@ private ITypeName GenericTypeArgumentsRule(Token genericTypeName, Token firstTok
-                    ParserStrings.EndSquareBracketExpectedAtEndOfAttribute);
-                commaOrRBracketToken = null;
+                    ParserStrings.EndSquareBracketExpectedAtEndOfAttribute,
+                    rBracketToken.Text);
```
The resource string `EndSquareBracketExpectedAtEndOfAttribute` doesn't take any args.

daxian-dbw replied:
Addressed.

daxian-dbw commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -1394,32 +1394,37 @@ private static bool IsInvocationConstraintSatisfied(OverloadCandidate overloadCa
+                MethodInformation methodInfo = methods[i];
 
-                if (method.method != null && method.method.DeclaringType.IsGenericTypeDefinition)
+                if ((methodInfo.method?.DeclaringType.IsGenericTypeDefinition == true)
```
Minor:
```suggestion
                if (methodInfo.method?.DeclaringType.IsGenericTypeDefinition == true
```

daxian-dbw commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -7889,6 +7895,31 @@ public MemberExpressionAst(IScriptExtent extent, ExpressionAst expression, Comma
             this.Member = member;
             SetParent(member);
             this.Static = @static;
+            this.GenericTypeArguments = new ReadOnlyCollection<ITypeName>(genericTypes ?? Array.Empty<ITypeName>());
```
Let's do the following to keep consistent with how `arguments` are handled by `InvokeMemberExpressionAst`:
```
  if (genericTypes != null && genericTypes.Count > 0)
  {
      this.Arguments = new ReadOnlyCollection<ITypeName>(genericTypes);
  }
```

daxian-dbw replied:
Addressed

daxian-dbw committed:
Address my review comments

daxian-dbw commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -7715,34 +7729,107 @@ private ExpressionAst MemberAccessRule(ExpressionAst targetExpr, Token operatorT
+                if (firstToken.Kind == TokenKind.Identifier || firstToken.Kind == TokenKind.LBracket)
+                {
+                    resyncIndex = -1;
+                    genericTypes = GenericTypeArgumentsRule(firstToken, unbracketedGenericArg: false, out rBracketToken);
```
Addressed

daxian-dbw commented on src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs:
```diff
@@ -2102,6 +2102,14 @@ private List<CompletionResult> GetResultForIdentifier(CompletionContext completi
+                && lastAst?.Find(a => a is MemberExpressionAst, searchNestedScriptBlocks: false) is not null)
+            {
+                result = CompletionCompleters.CompleteType(completionContext.TokenAtCursor.Text).ToList();
+            }
```
@vexx32 I've addressed all my other feedback, can you review my commits and address this last one?

vexx32 replied:
Appreciate it! I'll give them a look over and take care of this tonight. Thanks! 💖

iSazonov commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -7889,6 +7895,35 @@ public MemberExpressionAst(IScriptExtent extent, ExpressionAst expression, Comma
             SetParent(member);
             this.Static = @static;
+
+            if (genericTypes != null && genericTypes.Count > 0)
```
```suggestion
            if (genericTypes is not null && genericTypes.Count > 0)
```

iSazonov replied:
I am happy I contributed to this innovation feature! 😸

iSazonov commented on src/System.Management.Automation/resources/ExtendedTypeSystem.resx:
```diff
@@ -156,6 +156,12 @@
     <value>Cannot find an overload for "{0}" and the argument count: "{1}".</value>
   </data>
+  <data name="MethodGenericArgumentCountException" xml:space="preserve">
+    <value>Could not find a suitable generic method overload for "{0}" with "{1}" type parameters, and the argument count: "{2}".</value>
```
I see you follow the file format but I wonder why we put int in quotes.
```suggestion
    <value>Could not find a suitable generic method overload for "{0}" with "{1}" type parameters, and the argument count: {2}.</value>
```

vexx32 replied:
I didn't feel like this PR was the place to change that, but I don't disagree 😄

daxian-dbw replied:
Let's keep consistent with the existing error message.

vexx32 committed:
:white_check_mark: Add tests for tab completion

vexx32 committed:
Update src/System.Management.Automation/engine/parser/ast.cs

Co-authored-by: Ilya <darpa@yandex.ru>

vexx32 reviewed (commented):
Had a quick look over your edits and they look good to me, thanks for helping sort those out!

Also pushed a commit adding some tests for the tab completion, modelled after the existing tab completion tests for normal type token completions.

IISResetMe commented:
I realize this question is probably beyond the scope of this PR, but I feel it's still relevant:

Does it make sense to attempt coercion of enumerables passed as arguments to method parameters constrained by generic type parameters _when overload resolution is not otherwise ambiguous_?

Take the following example:

    $linq = [System.Linq.IEnumerable]
    
    # this fails because [object[]] doesn't implement [IEnumerable[int]] (not because of ambiguous overload resolution)
    $linq::Distinct[int](@(1,2,3,1,2,3))
    
    # this succeeds
    $linq::Distinct[int]([int[]]@(1,2,3,1,2,3))

vexx32 commented:
@IISResetMe I agree we should have that, and agree it's out of this scope... but I _would_ really love to have that. Because yeah, in some method calls it does feel like casts should not be necessary. Maybe would be good to file a follow up issue with that? I think it'll be easier to manage with the changes from this PR actually defining some of the types needed for the method call, so PS has less to infer from just the arguments.

daxian-dbw reviewed (approved):
LGTM

andschwa commented:
> We need @andschwa review too (for possible PSES breaking).

What are the concerns around breaking PSES?

iSazonov commented:
> > We need @andschwa review too (for possible PSES breaking).
> 
> What are the concerns around breaking PSES?

Changes in Parser.

daxian-dbw commented:
Thanks @vexx32 for your contribution!

> can we have better coercion semantics for downcasting to interfaces with covariant generic type parameters?

@IISResetMe Feel free to open an issue for discussion about this 😄
@andschwa I guess `EditorSyntax` would need update? If so, can you please open an issue tracking that?
@JamesWTruher Not sure how much this would affect PSSA. It would be great if you can help evaluate and open tracking issue as needed.

daxian-dbw commented:
Just noticed no documentation issue was filed. @vexx32 can you please open a documentation issue for this feature?

vexx32 commented:
Aye, will do!

oising commented:
w00000000. 

Congrats @vexx32

kborowinski commented:
@vexx32 Finally! This is awesome!
